### ISSUE TITLE
New pages require administrator permissions to upload

### ISSUE THREAD

**Issue by ryanboswell** [Smartling/wordpress-localization-plugin#163]
We recently discovered an issue where new pages can't be uploaded to Smartling for translation unless the user has the administrator role. We originally were running the 1.5.7 version of the plugin but updated to 1.5.11 to see if it was fixed and still see the problem. Unfortunately, I can't pinpoint exactly when this stopped working because we haven't added entirely new pages in quite a while.

Also, even for administrators, there's odd behavior where I can check a language and click the Upload button and when the page refreshes there is no indicator that the translation is in progress and it does not show up on the Translation Progress page/table. But if I immediately re-check the language and Upload again, it submits properly.

Once a new page is uploaded for at least one language, other users with Editor roles are able to submit to additional languages without a problem.

**Comment by paveldhq** [comment]
Hello @ryanboswell,
In order to send content for translation user role should have **smartling_connector_widget_cap** capability.
By default that capability is added only to **administrator** role.

**Comment by ryanboswell** [comment]
We do add that capability to the Editor role in our theme, and they're able to see the post meta box that administrators can. The problem shows up when they try to use it and upload the post for translation, after selecting any number of languages (in testing, it didn't matter if it was one, a random selection, or all languages) and submitting, nothing happens, regardless of how many times in a row they try.

I did do some digging and it appears that while the meta box is only shown to users with the `smartling_connector_widget_cap` capability, when it gets saved the `edit_post` capability is what actually gets checked before processing the post for translation ([starting here](https://github.com/Smartling/wordpress-localization-plugin/blob/e01d1d95b8a547f732cf4d65c9a9b51f75ab871d/inc/Smartling/WP/Controller/PostBasedWidgetControllerStd.php#L282-L284), I found the [capability here](https://github.com/Smartling/wordpress-localization-plugin/blob/e01d1d95b8a547f732cf4d65c9a9b51f75ab871d/inc/Smartling/WP/Controller/PostBasedWidgetControllerStd.php#L33)). That still doesn't explain what happened, but shows there's at least some kind of capability mismatch early in the process.

I wasn't able to dig much further than that, it was hard to follow the abstracted code to try and find out why there is different behavior for Admins vs. Editors despite both having the custom capability you defined and the `edit_post` capability that is actually checked on save.

**Comment by paveldhq** [comment]
@ryanboswell ,
We provided tests - added `smartling_connector_widget_cap` capability to `Editor` (standard) role and sent content (post) for translation. We didn't face described issue. Post was sent for translation and corresponding placeholder was created on another site.
To manage capabilities we used `User Role Editor` plugin.

Smartling-connector widget is displayed to users with `smartling_connector_widget_cap`, the `edit_post` capability is checked to find out if user is allowed to create posts and it is not related to widget.

Seems that you have customizations in `Editor` role (some capabilities are removed), please check standard `Editor` role with only added `smartling_connector_widget_cap` capability.

**Comment by ryanboswell** [comment]
Thanks for re-affirming things, I've gone back and done a more thorough battery of tests with user permission levels, and have confirmed that our Editor role does have the correct permissions.

What was getting in the way of that originally was uploads/downloads not always working the first time they were submitted, which I mentioned in the original issue I had to do as an administrator occasionally. It turns out the Editor I was testing with only tried submitting once, which led to the incorrect assumption that there was a user permission problem. On a second round of testing, we were able to get things working with 2-3 attempts on each page for an Editor user, which is consistent with the experience I was getting as an administrator.

The repeatable process we followed was to edit a page, select all languages, and click Upload. Usually on the first try nothing would happen, but on a subsequent try 95% of the time it would work and the submission would show as "In Progress", and the other 5% of the time it worked on the third attempt.

Unfortunately when I look through the logfiles that Smartling generates, there aren't any references to the post that fails. It only seems to log related entries when it submits successfully, so I can't narrow down where the problem happens.

**Comment by dimitrystd** [comment]
I could reproduce something similar but it was accidentally when i restored posts from backup. So before continuing the story, i would like to ask you, can you search in log "`just revision. Ignoring.`"? It should look like
```
[2017-08-16 20:37:13] 5994acf93504c DEBUG: Entering post save hook. post_id = '25', blog_id = '1'
[2017-08-16 20:37:13] 5994acf93504c DEBUG: Validation failed: post id = '25' just revision. Ignoring.
```

Now steps:
* Restore content by `wpcli import /tmp/wordpress-backup.xml --authors=skip`
* Open any post for editing http://localhost:8080/wp-admin/post.php?post=14&action=edit
* Submit the post for any locale. Page is reloaded but nothing happened
* In log i see messages above. Want to draw your attention that i submitted `post=14`, but my handler got `post=25`
* Try to submit once again (locale doesn't matter). Now everything is fine, post is submitted, and log looks like
```
[2017-08-16 20:37:58] 5994ad26c66ca DEBUG: Entering post save hook. post_id = '14', blog_id = '1'
[2017-08-16 20:37:58] 5994ad26c66ca DEBUG: Checking if content has changed for post blog=1, id=14 started.
[2017-08-16 20:37:58] 5994ad26c66ca DEBUG: No submissions found for post blog=1, id=14
[2017-08-16 20:37:58] 5994ad26c66ca DEBUG: Checking if content has changed for post blog=1, id=14 finished.
[2017-08-16 20:37:58] 5994ad26c66ca DEBUG: Validating post id = '14' saving
[2017-08-16 20:37:58] 5994ad26c66ca DEBUG: Starting saving submission: .....
```
 
I tried to do this by Admin and Editor. Both are affected, but for some reasons, Admin should always resubmit only once, Editor 2 or 3 times.
Of course, if i create a new post (by any role) then it's submitted from the 1st attempt.

Before we will dive into research this issue (why WP creates revision implicit), can you confirm that you have such log entries?

**Comment by ryanboswell** [comment]
I tried it again without an import step and was able to reproduce what you're seeing with the revision post logs, I had discounted those earlier because the IDs didn't match up and I assumed those were another user making changes elsewhere at the same time I was. So it does seem like that's the issue.

I do remember there is a [wp_is_post_revision](https://developer.wordpress.org/reference/functions/wp_is_post_revision/) function that can check for you and handily return the post ID of the original/parent post so that might help with ensuring you always deal with a 'real' post in the `save_post` hook.

**Comment by dimitrystd** [comment]
Don't remember why we decided to just ignore revision instead of trying to find an original post_id. We already use wp_is_post_revision, so i'm going to modify code and check both cases (post with and without revisions).

Meantime, can you share some details why your post may not have revisions? I can guess following cases:
* Revisions were disabled before (`define('WP_POST_REVISIONS', FALSE);` in wp-config.php) but later revisions were enabled.
* Some 3rd party modules that remove post revisions (compact DB).
* Post revisions were removed manually
* Custom code that disables revisions before post saving (like described [there](https://wordpress.stackexchange.com/a/220423))

**Comment by dimitrystd** [comment]
We made the fix b573b43b3348d35cdfd3675fe606bdc7396f01d6. Can you try this build https://www.dropbox.com/s/4fgkffe3vmkerwy/smartling-connector-1.6.1-dev.zip?dl=0 It works for me fine.

**Comment by ryanboswell** [comment]
Answering your first set of questions, we don't have revisions disabled and I don't believe we ever did, but we also started noticing the issue on newly created pages that didn't have many or any previous revisions yet.

As for the test build, we ran a few tests and it looks like that fix seems to have worked. We can get pages submitted on the first try now.

**Comment by dimitrystd** [comment]
Sounds good. Then close this ticket.

Do you need a release version? (the same code but with proper version, available on marketplace). Or let's wait in case something should be adjusted?

**Comment by ryanboswell** [comment]
For us to deploy the update, we'll need a tagged release since we keep all of our third party dependencies in sync with Composer.

**Comment by dimitrystd** [comment]
Released https://downloads.wordpress.org/plugin/smartling-connector.1.6.1.zip. Have good weekends.
